### DISCUSSION TITLE
JSON workflow file and workflow instances

### DISCUSSION THREAD

**Discussion by rmanjusha** [elsa-workflows/elsa-core#5433] [Q&A]
How to get workflow instance saved when loading JSON workflow definitions from file and executing workflow with WorkflowRunner.RunAsync.

**Comment by glime-ai** [comment]
To get the workflow instance saved when loading JSON workflow definitions from a file and executing the workflow with `WorkflowRunner.RunAsync`, you should follow these steps:

1. Deserialize the JSON file into a `Workflow` object using the `IWorkflowSerializer` service, not the `IActivitySerializer`. This is because workflows and activities are deserialized differently.

2. After deserialization, use the `IWorkflowRunner` service to execute the workflow. This service is responsible for running the workflow and can handle the execution of any `IActivity` object.

3. To save the workflow instance state at the beginning of its execution, you can manually create the instance with `IWorkflowInstanceStore` as you mentioned. However, if you want to avoid this manual step, you can use the `IWorkflowRuntime` service to start the workflow, which will automatically create and persist the workflow instance.

4. If you need to access the workflow instance ID immediately after starting the workflow, you can use the `IWorkflowRuntime` service to start the workflow and then access the `WorkflowInstance` object or its ID from the result returned by the `StartWorkflowAsync` method.

Here's a simplified example of how you might do this:

```csharp
var serializer = serviceProvider.GetRequiredService<IWorkflowSerializer>();
var workflowJson = await File.ReadAllTextAsync("path-to-your-json-file.json");
var workflow = serializer.Deserialize(workflowJson);

var workflowRunner = serviceProvider.GetRequiredService<IWorkflowRunner>();
var workflowInstance = await workflowRunner.RunAsync(workflow);
```

If you need to save the workflow instance state immediately, you can use the `IWorkflowInstanceStore` to persist the instance:

```csharp
var workflowInstanceStore = serviceProvider.GetRequiredService<IWorkflowInstanceStore>();
await workflowInstanceStore.SaveAsync(workflowInstance, cancellationToken);
```

Remember to handle exceptions and ensure that the workflow instance is correctly persisted in your data store.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).
